EZP-29206: MySQL deadlock errors happen when concurrently moving subtree

alongosz opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29206](https://jira.ez.no/browse/EZP-29206)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.7` and up.
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This PR presents one of the possible solutions for database deadlocks that occur when concurrently moving subtree. The solution was suggested by the Customer and improved by us.

The solution splits update-where-id-in-select query into two separate queries to avoid locks. While possible deadlocks are documented for InnoDB and solution is to set proper indices, there's no solution for the case when update and select operate on different tables.

Therefore obvious choice would be to split these operations. I was reluctant to do so because that introduces possible race condition between these select and update. However I haven't found so far real example where this would fail, so I'm presenting this as a possible solution.

**TODO**:
- [x] Split fetching content type Ids for update and update query into two separate queries
- [x] Introduce additional database indices.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [ ] Ask for Code Review.

gggeek commented:
What about adding a transaction around the 2 queries ? would it make sense ?

Without testing, logical cases where this might fail due to concurrency:
a) new node added that does not get its section updated
b) a node that gets moved out of the subtree but still gets updated

gggeek commented:
ps: I know that Oracle is not supported for now, but as a reminder for future evolutions, be aware that it has a limitation on the max number of elements that can be used in an IN condition

alongosz commented:
@gggeek this already is wrapped in a transaction on a higher level (persistence handler). I'll check if adding inner transaction is possible here, might be good for consistency and some weird direct gateway usages (which are not supported BTW :upside_down_face: )

gggeek commented on data/mysql/schema.sql:
```diff
@@ -625,7 +625,9 @@ CREATE TABLE `ezcontentobject` (
-  KEY `ezcontentobject_status` (`status`)
+  KEY `ezcontentobject_status` (`status`),
+  KEY `ezcontentobject_section` (`section_id`),
+  KEY `ezcontentobject_id_section` (`id`, `section_id`)
```
Possibly nitpick, but are you sure that you need this index ? And that you want to put the 'id' column first?

gggeek replied:
ps: as a matter of fact, the index ('section_id', 'id') is used for all queries where the index ('section_id') would be used, so you would need 1 less index.
To be tested of course, but I think that most databases are able to figures this out

alongosz replied:
> Possibly nitpick, but are you sure that you need this index ? And that you want to put the 'id' column first?

@gggeek TBH this is a little bit of InnoDB-dark-magic. it made a small difference in case of UPDATE-where-SELECT query that was used before. We tried to apply suggestions from [this doc](https://dev.mysql.com/doc/refman/8.0/en/innodb-deadlocks.html) to our case. However since I've separated these queries, I have no good point about keeping it.

I've updated PR - I've kept only `section_id` index as in general it might be useful.

gggeek commented on data/mysql/schema.sql:
```diff
@@ -764,7 +766,8 @@ CREATE TABLE `ezcontentobject_tree` (
   KEY `ezcontentobject_tree_path_ident` (`path_identification_string`(50)),
-  KEY `modified_subnode` (`modified_subnode`)
+  KEY `modified_subnode` (`modified_subnode`),
+  KEY `ezcontentobject_tree_contentobject_id_path_string` (`contentobject_id`, `path_string`)
```
same: you might want to put path_string 1st

alongosz replied:
> you might want to put path_string 1st

Indeed, on live data that might make a difference. Thanks, fixed.

gggeek commented:
@alongosz thx for the info.
If there's an outer transaction then it might not be worth adding a nested one, esp. considering that the side effects are not *extremely* severe... (and that nested transactions are a pain to debug anyway)

adamwojs commented on data/mysql/schema.sql:
```diff
@@ -625,7 +625,8 @@ CREATE TABLE `ezcontentobject` (
   KEY `ezcontentobject_lmask` (`language_mask`),
   KEY `ezcontentobject_owner` (`owner_id`),
   KEY `ezcontentobject_pub` (`published`),
-  KEY `ezcontentobject_status` (`status`)
```
Please apply these changes also to `/eZ/Publish/Core/Persistence/Legacy/Tests/_fixtures/schema.**.sql` scripts

gggeek commented:
As a side-note-cum-troll: going forward, if you want to integrate load testing into automated builds, to eg. avoid that perf regression are introduced or to verify the effects of newly added db indexes, I can suggest usage of ezloremipsumbundle to create a huge number of contents (and measure the time it takes), plus ezmigrationbundle to execute in a tight loop many content manipulation operations :-)

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/Location/Gateway/DoctrineDatabase.php:
```diff
@@ -1330,39 +1336,49 @@ public function removeElementFromTrash($id)
+        $selectContentIdsQuery = $this->connection->createQueryBuilder();
+        $selectContentIdsQuery
+            ->select($this->connection->quoteIdentifier('contentobject_id'))
+            ->from($this->connection->quoteIdentifier('ezcontentobject_tree'))
```
you can avoid having the quoting and make it more readable by doing something like:
```php
    ->select('t.contentobject_id')
    ->from('ezcontentobject_tree', 't')
```

same below

alongosz commented:
> Like we recently discussed on ezxmltext package, we can avoid having the quoting and make it more readable by doing something like:
> ```php
>    ->select('t.contentobject_id')
>    ->from('ezcontentobject_tree', 't')
> ```

@andrerom ok, changed. I keep quoting for consistency with other parts of the system - the ones that work on names that could potentially be DBMS keywords. In such case quoting is needed at least on `from` no matter what. But indeed here there's no such risk.

> same below

Removed quoting as well, but to clarify one thing: for updates such aliasing as the one above will not work. Not every DBMS supports that syntax (I only mention this because Doctrine allows it, which makes tests crash on sqlite :-) ).

@glye 
> This also adds a return value to setSectionForSubtree() where previously there was none despite doc, a nice bonus.

Yep, I can't look at such things, I have to fix that :grinning:  The same goes for long array syntax. 
Some people have to wash dirty dishes when they see it, some people have to fix unrelated code ;)
